ambry-account/src/main/java/com/github/ambry/account/AbstractAccountService.java

```diff
@@ -68,6 +77,84 @@ public Account getAccountById(short id) {
     return accountInfoMapRef.get().getAccounts();
   }
 
+  @Override
+  public Collection<Container> updateContainers(String accountName, Collection<Container> containers)
+      throws AccountServiceException {
+    checkOpen();
+    // input validation
+    if (accountName == null || accountName.isEmpty() || containers == null || containers.isEmpty()) {
+      throw new AccountServiceException("Account or container is null or empty", AccountServiceErrorCode.BadRequest);
+    }
+    Account account = getAccountByName(accountName);
+    if (account == null) {
+      logger.error("Account {} is not found", accountName);
+      throw new AccountServiceException("Account " + accountName + " is not found", AccountServiceErrorCode.NotFound);
+    }
+
+    List<Container> resolvedContainers = new ArrayList<>();
+    // create a hashmap to map the name to existing containers in account
+    Map<String, Container> existingContainersInAccount = new HashMap<>();
+    account.getAllContainers().forEach(c -> existingContainersInAccount.put(c.getName(), c));
+
+    // Generate container ids for new containers
+    short nextContainerId = account.getAllContainers()
+        .stream()
+        .map(Container::getId)
+        .max(Short::compareTo)
+        .map(maxId -> (short) (maxId + 1))
+        .orElse(config.containerIdStartNumber);
+
+    for (Container container : containers) {
+      if (container.getId() == Container.UNKNOWN_CONTAINER_ID) {
+        // new container
+        // make sure there is no conflicting container (conflicting means a container with same name but different attributes already exists).
+        Container existingContainer = existingContainersInAccount.get(container.getName());
+        if (existingContainer != null) {
+          if (existingContainer.isSameContainer(container)) {
+            // If an exactly same container already exists, ignore it. Adding same container multiple times is no-op.
```

Comment:
In current logic, if an exactly same container exists, the passed in container won't be added to final result. I wonder if we can still return the existing container in final result because its id is useful for closed source container creation logic.

For example: container creation involves two steps:  1. create container in ambry account service.  2. set up ACLs for this container.
If previous request succeeded on step 1 but failed on step 2 (i.e. ACL service is unavailable), the whole operation is not successful. We should allow user to retry, to be more accurate, retry the step 2 (which requires container id). So I suggest returning the existing container in result even for same container case. Thus, the caller can get container id and retry on setting up ACLs.

Reply:
This makes sense.  Existing containers are not included in the return list.